pkg/kubelet/nodestatus/manager.go

```diff
@@ -0,0 +1,762 @@
+	lastStatusReportTime time.Time
+
+	// handlers called during the tryUpdateNodeStatus cycle
+	nodeStatusFuncs []func(*v1.Node) error
+
+	// getNodeFunc is kubelet GetNode() function.
+	getNodeFunc func() (*v1.Node, error)
+
+	// updateRuntimeUpFunc is kubelet updateRuntimeUp function.
+	updateRuntimeUpFunc func()
+
+	// updatePodCIDRFunc is kubelet updatePodCIDR function.
+	updatePodCIDRFunc func(string) (bool, error)
+
+	// keepTerminatedPodVolumes is kubelet keepTerminatedPodVolumes.
+	keepTerminatedPodVolumes bool
+}
+
+// NewManager build Manager.
+func NewManager(
+	hostname string,
+	nodeName types.NodeName,
+	clock clock.Clock,
+	nodeLabels map[string]string,
+	registerNode bool,
+	kubeClient clientset.Interface,
+	heartbeatClient clientset.Interface,
+	providerID string,
+	externalCloudProvider bool,
+	containerManager cm.ContainerManager,
+	volumeManager volumemanager.VolumeManager,
+	cloud cloudprovider.Interface,
+	cloudResourceSyncManager cloudresource.SyncManager,
+	registerSchedulable bool,
+	registerWithTaints []api.Taint,
+	enableControllerAttachDetach bool,
+	onRepeatedHeartbeatFailure func(),
+	nodeStatusUpdateFrequency time.Duration,
+	nodeStatusReportFrequency time.Duration,
+	keepTerminatedPodVolumes bool,
+	getNodeFunc func() (*v1.Node, error),
+	updateRuntimeUpFunc func(),
+	updatePodCIDRFunc func(string) (bool, error),
+	nodeStatusFuncs []func(*v1.Node) error,
+) *Manager {
+	mgr := &Manager{
+		hostname:                     hostname,
+		nodeName:                     nodeName,
+		nodeLabels:                   nodeLabels,
+		clock:                        clock,
+		registerNode:                 registerNode,
+		kubeClient:                   kubeClient,
+		heartbeatClient:              heartbeatClient,
+		providerID:                   providerID,
+		externalCloudProvider:        externalCloudProvider,
+		containerManager:             containerManager,
+		volumeManager:                volumeManager,
+		cloud:                        cloud,
+		cloudResourceSyncManager:     cloudResourceSyncManager,
+		registerSchedulable:          registerSchedulable,
+		registerWithTaints:           registerWithTaints,
+		enableControllerAttachDetach: enableControllerAttachDetach,
+		onRepeatedHeartbeatFailure:   onRepeatedHeartbeatFailure,
+		nodeStatusUpdateFrequency:    nodeStatusUpdateFrequency,
+		nodeStatusReportFrequency:    nodeStatusReportFrequency,
+		keepTerminatedPodVolumes:     keepTerminatedPodVolumes,
+		getNodeFunc:                  getNodeFunc,
+		updateRuntimeUpFunc:          updateRuntimeUpFunc,
+		updatePodCIDRFunc:            updatePodCIDRFunc,
+		nodeStatusFuncs:              nodeStatusFuncs,
+	}
+	// check podCIDR here(current in kubelet).
+	return mgr
+}
+
+// Start sync node status.
+func (ns *Manager) Start() {
+	// Start syncing node status immediately, this may set up things the runtime needs to run.
+	go wait.Until(ns.syncNodeStatus, ns.nodeStatusUpdateFrequency, wait.NeverStop)
```

Comment:
maybe lets do:

```golang
go ns.fastStatusUpdateOnce()
wait.Until(ns.syncNodeStatus, ns.nodeStatusUpdateFrequency, wait.NeverStop)
```

It is generally preferable to have Start() called in its own goroutine, and for it to be blocking.

Reply:
done